File: designpatterns/java/src/iterator/Waitress.java
package iterator;

import java.util.Iterator;

public class Waitress {
    Menu breakfastMenu;
    Menu lunchMenu;
    Menu cafeMenu;

    public Waitress(Menu pancakeHouseMenu, Menu dinerMenu, Menu cafeMenu) {
        this.breakfastMenu = pancakeHouseMenu;
        this.lunchMenu = dinerMenu;
        this.cafeMenu = cafeMenu;
    }

    void printMenu() {
        Iterator pancakeIterator = breakfastMenu.createIterator();
        Iterator dinerIterator = lunchMenu.createIterator();
        Iterator cafeIterator = cafeMenu.createIterator();
        System.out.println("메뉴\n----\n아침 메뉴");
        printMenu(pancakeIterator);
        System.out.println("\n점심 메뉴");
        printMenu(dinerIterator);
        System.out.println("\n저녁 메뉴");
        printMenu(cafeIterator);
    }

    void printMenu(Iterator<MenuItem> iterator) {
       while(iterator.hasNext()) {
            MenuItem menuItem = iterator.next();
            System.out.print(menuItem.getName() + " ");
            System.out.println(menuItem.getPrice());
            System.out.println(menuItem.getDescription());
        }
    }


    void printVegetarianMenu(Iterator<MenuItem> iterator) {
        while(iterator.hasNext()) {
            MenuItem menuItem = iterator.next();
            if(isItemVegetarian(menuItem.getName(), iterator)) {
                System.out.print(menuItem.getName() + " ");
                System.out.println(menuItem.getPrice());
                System.out.println(menuItem.getDescription());
            }
        }

    }

    boolean isItemVegetarian (String name, Iterator<MenuItem> iterator) {
        while(iterator.hasNext()) {
            MenuItem menuItem = iterator.next();
            if(menuItem.getName().equals(name)) {
                return  menuItem.isVegetarian();
            }
        }
        return false;
    }
}

File: designpatterns/kotlin/designpatterns/src/main/kotlin/iterator/Waitress.kt
package iterator

class Waitress(
    private val pancakeHouseMenu: PancakeHouseMenu = PancakeHouseMenu(),
    private val dinerMenu: DinerMenu  = DinerMenu(),
) {

    fun printMenu() {
        val pancakeIterator = pancakeHouseMenu.creatIterator()
        val dinerIterator = dinerMenu.createIterator()

        println("아침 메뉴")
        printMenu(pancakeIterator)
        println("점심 메뉴")
        printMenu(dinerIterator)
    }

    private fun printMenu(iterator: Iterator<MenuItem>) {
        while (iterator.hasNext()) {
            val menuItem = iterator.next()
            println("${menuItem.name} ${menuItem.price} ${menuItem.description}")
        }
    }
}